### PR TITLE
Post-receive URLs

### PR DESCRIPTION
**Pull request by xdissent**
This PR adds post-receive URLs to the repository settings ala github. The POSTed payload is described here: http://help.github.com/post-receive-hooks/. I believe this is what #45 is asking for - I actually wrote it to trigger Jenkins builds.

### CODE REVIEW

**Comment by xdissent** [comment]
Oh it looks like 2 little commits made it in this PR that are unrelated - they just catch some exceptions that were giving me trouble when migrating from an older plugin version.

**Comment by xdissent** [comment]
I added support for different post-receive URL modes - `github` and `get`. I realized I needed an empty GET request but didn't want to ditch the github compatibility. _Now_ it works with jenkins =)

**Comment by kubitron** [comment]
I'll look at this.

**Comment by kalw** [comment]
Any news on this one ? =)

**Comment by kalw** [comment]
I tried the https://github.com/cjAdvertising/redmine_git_hosting.git version from xdissent
Everything works execpt when the post-receive-hook happens :

```
Processing ApplicationController#post_receive (for 192.168.128.91 at 2012-06-30 20:34:20) [POST]
  Parameters: {"asynch"=>"true", "action"=>"post_receive", "url"=>"http://localhost/githooks/post-receive", "refs"=>["13d94ca7c18a51b2d7662a1a6c19faa8e8a77b09,e8d68ff91b3bf3ae045c6f
079f841a35cc9fd6d1,refs/heads/master"], "clear_time"=>"1341088460", "controller"=>"gitolite_hooks", "debug"=>"false", "projectid"=>"tututu", "encoded_time"=>"7e81b6eec7da289b6b8f197
422e73385cb780a5c"}

SyntaxError (/usr/local/www/chiliproject/vendor/plugins/redmine_git_hosting/app/controllers/gitolite_hooks_controller.rb:148: syntax error, unexpected kDO_BLOCK, expecting '}'
                                payloads.each do |payload|
                                                ^
/usr/local/www/chiliproject/vendor/plugins/redmine_git_hosting/app/controllers/gitolite_hooks_controller.rb:156: syntax error, unexpected kEND, expecting '}'):
  thin (1.2.11) lib/thin/connection.rb:84:in `pre_process'
  thin (1.2.11) lib/thin/connection.rb:82:in `catch'
  thin (1.2.11) lib/thin/connection.rb:82:in `pre_process'
  thin (1.2.11) lib/thin/connection.rb:57:in `process'
  thin (1.2.11) lib/thin/connection.rb:42:in `receive_data'
  eventmachine (0.12.10) lib/eventmachine.rb:256:in `run_machine'
  eventmachine (0.12.10) lib/eventmachine.rb:256:in `run'
  thin (1.2.11) lib/thin/backends/base.rb:61:in `start'
  thin (1.2.11) lib/thin/server.rb:159:in `start'
  thin (1.2.11) lib/thin/controllers/controller.rb:86:in `start'
  thin (1.2.11) lib/thin/runner.rb:185:in `send'
  thin (1.2.11) lib/thin/runner.rb:185:in `run_command'
  thin (1.2.11) lib/thin/runner.rb:151:in `run!'
  thin (1.2.11) bin/thin:6
  /usr/local/bin/thin:23:in `load'
  /usr/local/bin/thin:23

Rendering /usr/local/www/chiliproject/public/500.html (500 Internal Server Error)
```

The tiny syntax error is solved by :

```
diff --git a/app/controllers/gitolite_hooks_controller.rb b/app/controllers/gitolite_hooks_controller.rb
index d378d0c..987d019 100644
--- a/app/controllers/gitolite_hooks_controller.rb
+++ b/app/controllers/gitolite_hooks_controller.rb
@@ -139,24 +139,24 @@ class GitoliteHooksController < ApplicationController
                        end

                        # Post to each post-receive URL
-                       @project.repository_post_receive_urls.all(:order => "active DESC, created_at ASC", :conditions => "active=1").each do |prurl|
+                       @project.repository_post_receive_urls.all(:order => "active DESC, created_at ASC", :conditions => "active=1").each {|prurl|
                                msg = "Posting #{payloads.length} post-receive payloads to #{prurl.url} ... "
                                GitHosting.logger.debug msg
                                output.write msg
                                output.flush
                                uri = URI(prurl.url)
-                               payloads.each do |payload|
+                               payloads.each {|payload|
                                        if prurl.mode == :github
-                                               res = Net::HTTP.post_form(uri, {"payload" => payload})
+                                               res = Net::HTTP.post_form(uri, {"payload" => payload.to_json})
                                        else
                                                res = Net::HTTP.get_response(uri)
                                        end
                                        output.write res.is_a?(Net::HTTPSuccess) ? "[success] " : "[failure] "
                                        output.flush
-                               end
+                               }
                                output.write "done\n"
                                output.flush
-                       end if @project.repository_post_receive_urls.any?
+                       } if @project.repository_post_receive_urls.any?

                        # Notify CIA
                        #Thread.abort_on_exception = true
```

From now everything works.

**Comment by kalw** [comment]
and i've changed the github like post form to json output .

**Comment by kubitron** [comment]
Ok.  This is a lot of code.

Can you summarize your basic approach?

**Comment by kalw** [comment]
@xdissent : we need you ! =)

**Comment by xdissent** [comment]
I basically just ripped off the repository mirror and CIA notification code that was already there and generalized it to work for post-recieve URLs. We've been using this for about a month with no issues so far.

**Comment by kubitron** [comment]
OK.  I take it that the "repository mirror" portion is an interface to specify one or more callback scripts?  I don't use the CIA notification code.  Is that already integrated with the existing post-receive script?  (I want to make sure that the receive-hook repair code that I wrote will still work).  

(Sorry if this sounds lazy -- haven't looked yet, just trying to get a feeling of what I am looking at to maximize the efficiency of understanding when I pull it down).

**Comment by kubitron** [comment]
Can you tell me about the problems that led to the first two commits here (the ones you call "accidentally included")?  What were you migrating from?  Were you unable to run "migrate_plugins"?  

I'm not even sure why the first one helped -- if you return "nil" to old_valuehash, then you would have a lot of nil['name'] references which would fail...

**Comment by kubitron** [comment]
Ok.  I've integrated most of these changes (except for the first two non-related commits). I also fixed the syntax error identified by @kalw. 

@xdissent and @kalw, please try pulling the master branch and see if it works as necessary. 

Also, @xdissent, if you could answer the question about the first 2 commits, that would be useful.  I didn't integrate them, but would be interested if there is a general problem that people will have upgrading from old versions...

**Comment by kubitron** [comment]
Not sure if either of you tried this code, but it turns out that it conflicted with a security patch.  You should repull to test.  (probably want to reset last few commits, since I removed commit and forced --- bad, I know, but didn't want that commit in the history).

**Comment by kalw** [comment]
Hi there,

The only code i tried is the whole plugin from https://github.com/cjAdvertising/redmine_git_hosting.git .
Not sure to be helpfull on this one execpt for monkey testing =)

**Comment by xdissent** [comment]
Hi! Sorry I've been pretty busy but I did get to try this last night and all seemed to be well. What was it conflicting with? Is master supposed to be fixed at this point?

Oh, and those 2 commits that were random fixes - I'll send a PR for 1, but I think the other was specific to my situation because I was running a super old fork of the plugin and the migrations weren't working. The one that's still relevant prevents rails from bailing when it tries to include the routes and the database doesn't yet exist. Since rails loads the routes immediately, rake tasks fail if you didn't run `db:migrate` before adding the plugin.

**Comment by kubitron** [comment]
I pushed fix yesterday, so you got the right code.

You might try pulling my "testing" branch.  It has a couple more commits, one of which does a better job of dealing with validation errors when setting up a new post-receive URLs (I needed to make validation errors appear on popups for the update to the repository mirrors functionality. Added it to the post-receive URLs as well).  I am waiting on someone else before moving the testing branch to master, but this code is pretty stable.

Please register an actual bug if you think that there is one during initial migration.

**Comment by kubitron** [comment]
@kalw you should try this version of the plugin, since I believe that it is the more maintained version.  I thought that you were querying trying to get the merge to happen.

**Comment by kalw** [comment]
@john: i was querying to merge the functionality =)

On Tue, Jul 17, 2012 at 5:15 PM, John Kubiatowicz
reply@reply.github.com
wrote:

**Comment by kalw** [comment]
Sorry for late reply.

Tested the master branch. works like a charm.

Thanks @kubitron & @xdissent .

**Comment by kalw** [comment]
i might have a comment regarding the payload content :

for example : 

```
"53991580696e7537682ccf9796d04f31b36dc291", "message": "coincoin", "modified": [], "removed": [], "timestamp": "2012-07-23T09:58:05+00:00", "url": "https://localhost:3000/projects/titi/repository/revisions/53991580696e7537682ccf9796d04f31b36dc291" } ], "ref": "refs/heads/master", "repository": { "description": "", "fork": false, "forks": 0, "homepage": "", "name": "titi", "open_issues": 0, "owner": { "email": "chiliproject@example.net", "name": "Projects" }, "private": false, "url": "https://localhost:3000/projects/titi/repository", "watchers": 0 } }
```

both "url" contents seems taken from server infos rather than plugin config.

I'm not quite a ruby expert, but i'll try to make a patch if someone feels the need.

**Comment by kubitron** [comment]
@kalw -- I believe that the url is supposed to refer to the repository and/or commit URL within redmine.  This looks right to me... Isn't it?

**Comment by kubitron** [comment]
@kalw -- Also, can you try pulling from my testing branch?  It should still work the same way, but has some changed internals since I used this code for my extended mirroring patch (and the form validation works properly).

**Comment by kalw** [comment]
@kubitron : thanks . i will as soon as found something revelent.

**Comment by kubitron** [comment]
Thought I'd let you guys know that I fixed a bug in the post-receive hook under Ruby 1.8.7 that would have prevented multiple refs from being forwarded (when you commit more than one branch at a time).

I've updated the master branch with these fixes.

**Comment by kalw** [comment]
Cool thx
